tests/bsim/bluetooth/tester/src/btp.c

```diff
@@ -0,0 +1,177 @@
+	btp_buffer = btp_cb(btp_buffer, &offset);
+	TEST_ASSERT(offset == 0U);
+
+	wait_for_response(cmd_hdr);
+}
+
+void bsim_btp_register_tester(uint8_t *buffer, size_t len, uart_pipe_recv_cb cb)
+{
+	TEST_ASSERT(cb != NULL);
+	TEST_ASSERT(buffer != NULL);
+	TEST_ASSERT(len > sizeof(struct btp_hdr), "len %zu", len);
+	TEST_ASSERT(len <= BTP_MTU, "len %zu > %d", len, BTP_MTU);
+
+	btp_buffer = buffer;
+	btp_buffer_len = len;
+	btp_cb = cb;
+
+	LOG_INF("btp registered with %p %zu %p", buffer, len, cb);
+}
+
+void bsim_btp_send_to_bsim(const uint8_t *data, size_t len)
+{
+	static struct net_buf *buf;
+	bool full_evt;
+
+	TEST_ASSERT(len >= 0);
+	TEST_ASSERT(data != NULL);
+	TEST_ASSERT(len <= BTP_MTU - sizeof(struct btp_hdr), "len %zu < %d", len, BTP_MTU);
+	TEST_ASSERT(buf != NULL || (buf == NULL && len == sizeof(struct btp_hdr)), "len %zu buf %p",
+		    len, buf);
+
+	/* tester_send_with_index always sends the header first, and then any additional information
+	 * afterwards. We thus need to reassemble the full event.
+	 */
+	if (buf == NULL) {
+		const struct btp_hdr *hdr = (const struct btp_hdr *)data;
+
+		LOG_DBG("hdr service %u and opcode %u", hdr->service, hdr->opcode);
+		TEST_ASSERT(hdr->opcode != BTP_STATUS, "hdr service %u", hdr->service);
+
+		if (hdr->opcode < BTP_EVENT_OPCODE) {
+			buf = net_buf_alloc(&btp_rsp_pool, K_NO_WAIT);
+			TEST_ASSERT(buf != NULL);
+		} else {
+			buf = net_buf_alloc(&btp_evt_pool, K_NO_WAIT);
+			if (buf == NULL) {
+				/* Discard the oldest event */
+				buf = k_fifo_get(&btp_evt_fifo, K_NO_WAIT);
+				TEST_ASSERT(buf != NULL);
+				net_buf_unref(buf);
+
+				buf = net_buf_alloc(&btp_evt_pool, K_NO_WAIT);
+				TEST_ASSERT(buf != NULL);
+			}
+		}
+
+		full_evt = hdr->len == 0; /* we are not awaiting additional data if hdr->len == 0*/
+	} else {
+		/* If buf != NULL then we have received the hdr and this is the additional data for
+		 * the event/response. We always assume that we receive the rest here
+		 */
+		const struct btp_hdr *hdr = (const struct btp_hdr *)buf->data;
+
+		TEST_ASSERT(buf != NULL && len == hdr->len, "len %zu hdr len %u", len, hdr->len);
+		full_evt = true;
+	}
+
+	net_buf_add_mem(buf, data, len);
+
+	if (full_evt) {
+		const struct btp_hdr *hdr = (const struct btp_hdr *)buf->data;
+
+		if (hdr->opcode < BTP_EVENT_OPCODE) {
+			k_fifo_put(&btp_rsp_fifo, buf);
+		} else {
+			k_fifo_put(&btp_evt_fifo, buf);
+		}
+
+		buf = NULL;
+	}
+}
+
+void bsim_btp_wait_for_evt(uint8_t service, uint8_t opcode, struct net_buf **out_buf)
+{
+	LOG_DBG("Waiting for evt with service %u and opcode %u", service, opcode);
+
+	while (true) {
+		const struct btp_hdr *hdr;
+		struct net_buf *buf;
+
+		buf = k_fifo_get(&btp_evt_fifo, K_FOREVER);
+		TEST_ASSERT(buf != NULL);
+
+		LOG_HEXDUMP_DBG(buf->data, buf->len, "evt");
+		hdr = net_buf_pull_mem(buf, sizeof(struct btp_hdr));
+
+		/* TODO: Verify length of event based on the service and opcode */
```

Comment:
did you forget this todo?

Reply:
Will add validation of events and responses in this PR :)